src/utils.jl

```diff
@@ -9,6 +9,29 @@ function check_nrows(X::AbstractMatrix, y::AbstractVecOrMat)::Nothing
     throw(DimensionMismatch("`X` and `y` must have the same number of rows."))
 end
 
+function verify_or_construct_gramian(glr, X, y, data)
+    check_nrows(X, y)
+    isnothing(data) && return (; XX = X'X, Xy = X'y, n = length(y))
+
+    !all(hasproperty.(Ref(data), (:XX, :Xy, :n))) && throw(ArgumentError("data must contain XX, Xy, n"))
+    size(data.XX, 1) != size(data.Xy, 1) && throw(DimensionMismatch("`XX` and Xy` must have the same number of rows."))
+    !issymmetric(data.XX) && throw(ArgumentError("Input `XX` must be symmetric"))
+
+    c = getc(glr, data.Xy)
+    !iszero(c) && throw(ArgumentError("Categorical loss not supported with Gramian kernel"))
+    glr.fit_intercept && throw(ArgumentError("Intercept not supported with Gramian kernel"))
+
+    if any(.!iszero.((X, y)))
+        all((
+            isapprox(X'X, data.XX; rtol=1e-5),
+            isapprox(X'y, data.Xy; rtol=1e-5),
+            length(y) == data.n
+        )) || throw(ArgumentError("Inputs `X` and `y` do not match inputs `XX` and `Xy`."))
+    end
```

Comment:
I might be misunderstanding this check, if a user provides `data` then is it the case that `XX` and `Xy` are (re)computed to check that `data` is correct?

Reply:
yes, that was the intention. if `data` only is provided, `XX, Xy` used as given. if `X, y` only provided, then `XX, Xy` recomputed and used. if both `data` and `X, y` provided, then `XX, Xy` are recomputed to catch the user from making silly errors, but the original values in `data` are used

note the the added method `fit(glr::GLR; kwargs...) = fit(glr, zeros((0,0)), zeros((0,)); kwargs...)` to allow calling without `X,y` at all. I changed the `iszero` check to `isempty` to make it clear those are dummy variables

Reply:
> if both data and X, y provided, then XX, Xy are recomputed to catch the user from making silly errors, but the original values in data are used

ok this is what I hadn't understood, thanks for the clarification